### ISSUE TITLE
[Clang] False positive when checking `-Warray-bounds` for C-arrays

### ISSUE THREAD

**Issue by JMazurkiewicz** [llvm/llvm-project#61025]
Repro:

Checking if it is fine to access array before doing it, does not prevent `-Warray-bounds` warning.

```c++
template <class T, unsigned N>
constexpr auto size(const T (&)[N]) {
    return N;
}

template<class R>
constexpr bool test(const R& r) {
    if(size(r) > 1) {
        (void) (r[1] + 1); // if we remove `+ 1`, the error will disappear
    }
    return true;
}

constexpr int data[] = {10};
static_assert(test(data));
```

Result:

```console
<source>:9:17: warning: array index 1 is past the end of the array (that has type 'const int[1]') [-Warray-bounds]
        (void) (r[1] + 1); // if we remove `+ 1`, the error will disappear
                ^ ~
<source>:15:15: note: in instantiation of function template specialization 'test<int[1]>' requested here
static_assert(test(data));
              ^
<source>:7:21: note: array 'r' declared here
constexpr bool test(const R& r) {
                    ^
```

Compiler explorer: https://godbolt.org/z/d7b55PTb8
